File: RotateImage/RotateImage.py
class Solution(object):
    def rotate(self, matrix):
        """
        :type matrix: List[List[int]]
        :rtype: void Do not return anything, modify matrix in-place instead.
        """
        for i in range(len(matrix) / 2):
            last = len(matrix) - 1 - i
            for j in range(i, last):
                offset = j - i
                top = matrix[i][j] # save top
                # left -> top
                matrix[i][j] = matrix[last - offset][i]
                # bottom -> left
                matrix[last - offset][i] = matrix[last][last - offset]
                # right -> bottom
                matrix[last][last - offset] = matrix[j][last]
                # top -> right
                matrix[j][last] = top

File: RotateImage/RotateImage.cpp
class Solution {
public:
    void rotate(vector<vector<int>>& matrix) {
        int index = matrix.size() / 2;
        for(int i = 0; i < index; i++) {
        	int last = matrix.size() - 1 - i;
        	for(int j = i; j < last; j++) {
        		int offset = j - i;
        		int top = matrix[i][j];
        		matrix[i][j] = matrix[last - offset][i];
        		matrix[last - offset][i] = matrix[last][last - offset];
        		matrix[last][last - offset] = matrix[j][last];
        		matrix[j][last] = top;
        	}
        }
    }
};